Document file realm and roles in the ES spec (#2703)
elastic/cloud-on-k8s: docs/orchestrating-elastic-stack-applications/elasticsearch/users-and-roles.asciidoc

:parent_page_id: elasticsearch-specification
:page_id: users-and-roles
ifdef::env-github[]
****
link:https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-{parent_page_id}.html#k8s-{page_id}[View this document on the Elastic website]
****
endif::[]
[id="{p}-{page_id}"]
= Users and roles

== Default elastic user

> A default user named `elastic` is created automatically, and is assigned the `superuser` role.

Reviewer: Maybe it's obvious, but would it be worth to remind here *when* the default user is created?

Its password can be retrieved in a Kubernetes secret, whose name
is based on the Elasticsearch resource name: `<elasticsearch-name>-es-elastic-user`.

For example, the password of the `elastic` user for an Elasticsearch cluster named `quickstart` can be retrieved with:

[source,sh]
----

> kubectl get secret quickstart-es-elastic-user -o=jsonpath='{.data.elastic}' | base64 --decode

Reviewer 2: Not that it matters much but `kubectl get secret quickstart-es-elastic-user -o go-template='{{.data.elastic | base64decode}}' would work without `base64` installed on the users machine

Author: Good point, also changing this in the quickstart.

----

== Creating custom users

=== Native realm

> Custom users can be created in the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/native-realm.html[Elasticsearch native realm] using link:https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api.html#security-user-apis[Elasticsearch user management APIs].

Reviewer: ```suggestion
You can create custom users in the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/native-realm.html[Elasticsearch native realm] using link:https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api.html#security-user-apis[Elasticsearch user management APIs].
```
We can try to make all the sentences more actionable and user-centric

=== File realm

> Custom users can also be created by providing the desired link:https://www.elastic.co/guide/en/elasticsearch/reference/current/file-realm.html[file realm content]

Reviewer: ```suggestion
You can also create custom users by providing the desired link:https://www.elastic.co/guide/en/elasticsearch/reference/current/file-realm.html[file realm content]
```

in Kubernetes secrets, referenced in the Elasticsearch resource.

> [source,yaml]

Reviewer 2: ```suggestion
[source,yaml,subs="attributes"]
```

----
apiVersion: elasticsearch.k8s.elastic.co/{eck_crd_version}
kind: Elasticsearch
metadata:
  name: elasticsearch-sample
spec:
  version: {version}
  auth:
    fileRealm:
    - secretName: my-filerealm-secret-1
    - secretName: my-filerealm-secret-2
  nodeSets:
  - name: default
    count: 1
----

> Several secrets can be referenced in the Elasticsearch specification.

Reviewer: ```suggestion
You can reference several secrets in the Elasticsearch specification.
```

> You can reference several secrets in the Elasticsearch specification.

Reviewer 3: Nit: no need for a newline between these sentences

Author: I tend to think it makes reading the "source" asciidoc file easier, and is not rendered in the final output.

ECK aggregates their content into a single secret, mounted in every Elasticsearch Pod.

> Referenced secrets may be composed of 2 entries:

Reviewer 3: Maybe worth clarifying that you can provide either or both?

```suggestion
Referenced secrets may be composed of 2 entries. You can provide either entry or both entries in each secret:
```

- `users`: content of the `users` file. It specifies user names and password hashes, as described in the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/file-realm.html[file realm documentation].
- `users_roles`: content of the `users_roles` file. It associates each role to a list of users, as described in the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/file-realm.html[file realm documentation].

> If multiple users with the same name are specified in more than one secret, the last one takes precedence.

Reviewer: ```suggestion
If you specify multiple users with the same name in more than one secret, the last one takes precedence.
```

> If multiple roles are specified in more than one secret, a single entry per role will be derived from the concatenation of its corresponding users from all secrets.

Reviewer: ```suggestion
If you specify multiple roles in more than one secret, a single entry per role is derived from the concatenation of its corresponding users from all secrets.
```

The same comment applies in similar sentences below.

> If you specify multiple roles with the same name in more than one secret, a single entry per role is be derived from the concatenation of its corresponding users from all secrets.

Reviewer 2: ```suggestion
If you specify multiple roles with the same name in more than one secret, a single entry per role is derived from the concatenation of its corresponding users from all secrets.
```

The following Secret specifies three users and their respective roles:

[source,yaml]
----
kind: Secret
apiVersion: v1
metadata:
  name: my-filerealm-secret
stringData:
  users: |-
    rdeniro:$2a$10$BBJ/ILiyJ1eBTYoRKxkqbuDEdYECplvxnqQ47uiowE7yGqvCEgj9W
    alpacino:$2a$10$cNwHnElYiMYZ/T3K4PvzGeJ1KbpXZp2PfoQD.gfaVdImnHOwIuBKS
    jacknich:{PBKDF2}50000$z1CLJt0MEFjkIK5iEfgvfnA6xq7lF25uasspsTKSo5Q=$XxCVLbaKDimOdyWgLCLJiyoiWpA/XDMe/xtVgn1r5Sg=
  users_roles: |-
    admin:rdeniro
    power_user:alpacino,jacknich
    user:jacknich
----

> The content of both `users` and `users_roles` can be populated using the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/users-command.html[elasticsearch-users] tool.

Reviewer: ```suggestion
You can populate the content of both `users` and `users_roles` using the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/users-command.html[elasticsearch-users] tool.
```

For example, invoking the tool in a Docker container:

[source,sh]
----
# create a folder with the 2 files
mkdir filerealm
touch filerealm/users filerealm/users_roles

# create user 'myuser' with role 'monitoring_user'
docker run \
    -v ${PWD}/filerealm:/usr/share/elasticsearch/config \
    docker.elastic.co/elasticsearch/elasticsearch:{version} \
    bin/elasticsearch-users useradd myuser -p mypassword -r monitoring_user

# create a Kubernetes secret with the file realm content

> kubectl create secret generic my-file-realm-secret --from-file filerealm

Author: This bash snippet is not great, but better than nothing, and better than downloading the release and installing Java IMHO.

Reviewer 2: 👍

----

== Creating custom roles

link:https://www.elastic.co/guide/en/elasticsearch/reference/current/defining-roles.html[Roles] can be specified using the
link:https://www.elastic.co/guide/en/elasticsearch/reference/current/defining-roles.html#roles-management-api[Role management API],
or the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/defining-roles.html#roles-management-ui[Role management UI in Kibana].

Additionally, link:https://www.elastic.co/guide/en/elasticsearch/reference/current/defining-roles.html#roles-management-file[file-based role management] can be achieved by referencing Kubernetes secrets containing the roles specification.

> [source,yaml]

Reviewer 2: ```suggestion
[source,yaml,subs="attributes"]
```

----
apiVersion: elasticsearch.k8s.elastic.co/{eck_crd_version}
kind: Elasticsearch
metadata:
  name: elasticsearch-sample
spec:
  version: {version}
  auth:
    roles:
    - secretName: my-roles-secret-1
    - secretName: my-roles-secret-2
  nodeSets:
  - name: default
    count: 1
----

Several secrets can be referenced in the Elasticsearch specification.
ECK aggregates their content into a single secret, mounted in every Elasticsearch Pod.

> Each secret must have a `roles.yml` entry, containing the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/defining-roles.html#roles-management-file[roles configuration].

Reviewer 3: "roles configuration" sounds wrong (because it is the configuration of the roles) but I'm not sure how it could be improved. "roles' configuration" also doesn't seem quite right. Plural "roles" is correct because you can define multiple roles in the config. I'm fine with it as is but wanted to raise it in case someone had a better idea.

Author: "roles specification" instead of "roles configuration"?

Author: Or maybe "role definition" since the ES doc is called "defining roles".

Reviewer 2: +1 for "roles definition"
